src/ORM/Search/BasicSearchContext.php

```diff
@@ -0,0 +1,186 @@
+use SilverStripe\ORM\Filters\SearchFilter;
+use SilverStripe\ORM\Limitable;
+use SilverStripe\ORM\Sortable;
+
+/**
+ * A SearchContext that can be used with non-ORM data.
+ * This class isn't guaranteed to respect the full searchable fields spec defined on DataObject classes.
+ */
+class BasicSearchContext extends SearchContext
+{
+    use Configurable;
+
+    /**
+     * Name of the field which, if included in search forms passed to this object, will be used
+     * to search across all searchable fields.
+     */
+    private static $general_search_field_name = 'q';
+
+    /**
+     * Returns a list which has been limited, sorted, and filtered by the given parameters.
+     *
+     * @param array $searchParams Map of search criteria, mostly taken from $_REQUEST.
+     *  If a filter is applied to a relationship in dot notation,
+     *  the parameter name should have the dots replaced with double underscores,
+     *  for example "Comments__Name" instead of the filter name "Comments.Name".
+     * @param array|bool|string $sort Field to sort on.
+     * @param array|null|string $limit
+     * @param Filterable&Sortable&Limitable $existingQuery
+     */
+    public function getQuery($searchParams, $sort = false, $limit = false, $existingQuery = null): Filterable&Sortable&Limitable
+    {
+        if (!$existingQuery || !($existingQuery instanceof Filterable) || !($existingQuery instanceof Sortable) || !($existingQuery instanceof Limitable)) {
+            throw new InvalidArgumentException('getQuery requires a pre-existing filterable/sortable/limitable list to be passed as $existingQuery.');
+        }
+
+        if ((count(func_get_args()) >= 3) && (!in_array(gettype($limit), ['array', 'NULL', 'string']))) {
+            Deprecation::notice(
+                '5.1.0',
+                '$limit should be type of array|string|null'
+            );
+            $limit = null;
+        }
+
+        $searchParams = $this->applySearchFilters($this->normaliseSearchParams($searchParams));
+        $result = $this->applyGeneralSearchField($searchParams, $existingQuery);
+
+        // Filter the list by the requested filters.
+        if (!empty($searchParams)) {
+            $result = $result->filter($searchParams);
+        }
+
+        // Only sort if a sort value is provided - sort by "false" just means use the existing sort.
+        if ($sort) {
+            $result = $result->sort($sort);
+        }
+
+        // Limit must be last so that ArrayList results don't have an applied limit before they can be filtered/sorted.
+        $result = $result->limit($limit);
+
+        return $result;
+    }
+
+    private function normaliseSearchParams(array $searchParams): array
+    {
+        $normalised = [];
+        foreach ($searchParams as $field => $searchTerm) {
+            $normalised[str_replace('__', '.', $field)] = $searchTerm;
+        }
+        return array_filter($normalised, [$this, 'clearEmptySearchFields']);
```

Comment:
We are looping through our search params twice with the current implementation

```suggestion
        $normalised = [];
        foreach ($searchParams as $field => $searchTerm) {
            if ($this->clearEmptySearchFields($searchTerm)) {
                $normalised[str_replace('__', '.', $field)] = $searchTerm;
            }
        }
        return $normalised;
```

Not overly fuss about this.

Reply:
Done